File: src/test/java/com/binance/api/examples/CandlesticksCacheExample.java
package com.binance.api.examples;

import com.binance.api.client.BinanceApiAsyncRestClient;
import com.binance.api.client.BinanceApiClientFactory;
import com.binance.api.client.BinanceApiWebSocketClient;
import com.binance.api.client.domain.market.Candlestick;
import com.binance.api.client.domain.market.CandlestickInterval;

import java.util.Map;
import java.util.TreeMap;
import java.util.concurrent.CompletableFuture;

import static java.util.Optional.empty;

/**
 * Illustrates how to use the klines/candlesticks event stream to create a local cache of bids/asks for a symbol.
 */
public class CandlesticksCacheExample {

  public CandlesticksCacheExample(String symbol, CandlestickInterval interval) {
    BinanceApiClientFactory factory = BinanceApiClientFactory.newInstance();
    BinanceApiAsyncRestClient client = factory.newAsyncRestClient();
    BinanceApiWebSocketClient webSocketClient = factory.newWebSocketClient();

    initializeCandlestickCache(client, symbol, interval).thenAccept(
        cache -> startCandlestickEventStreaming(webSocketClient, cache, symbol, interval));
  }

  /**
   * Initializes the candlestick cache by using the REST API.
   */
  private CompletableFuture<Map<Long, Candlestick>> initializeCandlestickCache(BinanceApiAsyncRestClient client, String symbol, CandlestickInterval interval) {
    return client.getCandlestickBars(symbol.toUpperCase(), interval, empty(), empty(), empty()).thenApply(
        candlestickBars -> {
          Map<Long, Candlestick> candlesticksCache = new TreeMap<>();
          for (Candlestick candlestickBar : candlestickBars) {
            candlesticksCache.put(candlestickBar.openTime, candlestickBar);
          }
          return candlesticksCache;
        });
  }

  /**
   * Begins streaming of depth events.
   */
  private void startCandlestickEventStreaming(BinanceApiWebSocketClient client,
                                              Map<Long, Candlestick> candlesticksCache,
                                              String symbol,
                                              CandlestickInterval interval) {
    client.onCandlestickEvent(symbol.toLowerCase(), interval, response -> {
      candlesticksCache.put(response.openTime, response);
      System.out.println(response);
    });
  }

  public static void main(String[] args) {
    new CandlesticksCacheExample("ETHBTC", CandlestickInterval.ONE_MINUTE);
  }
}

File: src/test/java/com/binance/api/examples/CandlesticksCacheExample.scala
package com.binance.api.examples

import java.util

import com.binance.api.client.domain.CandlestickBase
import com.binance.api.client.domain.event.CandlestickEvent
import com.binance.api.client.domain.market.CandlestickInterval
import com.binance.api.client.{BinanceApiAsyncRestClient, BinanceApiClientFactory, BinanceApiWebSocketClient}

import scala.concurrent.Future
import scala.concurrent.ExecutionContext.Implicits.global

/**
  * Illustrates how to use the klines/candlesticks event stream to create a local cache of bids/asks for a symbol.
  */
object CandlesticksCacheExample extends App {
  type Cache = util.TreeMap[Long, CandlestickBase]

  val symbol          = "ETHBTC"
  val interval        = CandlestickInterval.ONE_MINUTE
  val factory         = new BinanceApiClientFactory("", "")
  val client          = factory.newAsyncRestClient
  val webSocketClient = factory.newWebSocketClient

  /**
    * Initializes the candlestick cache by using the REST API.
    */
  def initializeCandlestickCache(client:   BinanceApiAsyncRestClient,
                                 symbol:   String,
                                 interval: CandlestickInterval): Future[Cache] =
    client.getCandlestickBars(symbol.toUpperCase, interval).map { candlestickBars =>
      val candlesticksCache = new Cache
      for (candlestickBar <- candlestickBars) {
        candlesticksCache.put(candlestickBar.openTime, candlestickBar)
      }
      candlesticksCache
    }

  /**
    * Begins streaming of depth events.
    */
  def startCandlestickEventStreaming(client:            BinanceApiWebSocketClient,
                                     candlesticksCache: Cache,
                                     symbol:            String,
                                     interval:          CandlestickInterval) =
    client.onCandlestickEvent(symbol.toLowerCase, interval) { (response: CandlestickEvent) =>
      candlesticksCache.put(response.candlestick.openTime, response.candlestick)
      System.out.println(response)
    }

  initializeCandlestickCache(client, symbol, interval).foreach(
    cache => startCandlestickEventStreaming(webSocketClient, cache, symbol, interval)
  )
}
